sdks/go/pkg/beam/core/runtime/exec/timers.go

```diff
@@ -0,0 +1,123 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exec
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/core/graph/coder"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/core/graph/mtime"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/core/timers"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/core/typex"
+)
+
+// UserTimerAdapter provides a timer provider to be used for manipulating timers.
+type UserTimerAdapter interface {
+	NewTimerProvider(ctx context.Context, manager DataManager, inputTimestamp typex.EventTime, windows []typex.Window, element *MainInput) (timerProvider, error)
+}
+
+type userTimerAdapter struct {
+	sID            StreamID
+	timerIDToCoder map[string]*coder.Coder
+}
+
+// NewUserTimerAdapter returns a user timer adapter for the given StreamID and timer coder.
+func NewUserTimerAdapter(sID StreamID, c *coder.Coder, timerCoders map[string]*coder.Coder) UserTimerAdapter {
+	if !coder.IsW(c) {
+		panic(fmt.Sprintf("expected WV coder for user timer %v: %v", sID, c))
+	}
+
+	return &userTimerAdapter{sID: sID, timerIDToCoder: timerCoders}
+}
+
+// NewTimerProvider creates and returns a timer provider to set/clear timers.
+func (u *userTimerAdapter) NewTimerProvider(ctx context.Context, manager DataManager, inputTs typex.EventTime, w []typex.Window, element *MainInput) (timerProvider, error) {
+	userKey := &FullValue{Elm: element.Key.Elm}
+	tp := timerProvider{
+		ctx:             ctx,
+		tm:              manager,
+		userKey:         userKey,
+		inputTimestamp:  inputTs,
+		sID:             u.sID,
+		window:          w,
+		writersByFamily: make(map[string]io.Writer),
+		codersByFamily:  u.timerIDToCoder,
```

Comment:
4. Per this bit of jumping around, since it's always going to be the same coder for all families in the same DoFn, we can pre-make the Encoder here too, so it's not created on every call to Set.

Reply:
Done